text/3176-cargo-multi-dep-artifacts.md

```diff
@@ -0,0 +1,184 @@
+```
+
+Cargo will make the binaries from each of these artifacts available under the
+specified name. For instance, in this example, binaries from `example` built
+for `riscv64imac_unknown_none_elf` will appear in the directory specified by
+the environment variable `CARGO_BIN_DIR_EXAMPLE_RISCV`, while binaries from
+`example` built for `aarch64-unknown-none` will appear in the directory
+specified by `CARGO_BIN_DIR_EXAMPLE_ARM`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Cargo allows specifying multiple dependencies on the same crate, as long as all
+such dependencies unify to the same version with the same features, and have
+different dependency names specified. Cargo will make the dependency available
+under each specified name.
+
+Multiple artifact dependencies on the same crate may have different `target`
+fields. In this case, cargo will build the dependency for each specified
+`target`, and make each build available via the corresponding dependency name.
+
+Cargo provides your crate with the standard set of environment variables for
+each artifact dependency: `CARGO_<ARTIFACT-TYPE>_DIR_<DEP>` for the directory
+containing the artifacts (e.g.  `CARGO_BIN_DIR_EXAMPLE`) and
+`CARGO_<ARTIFACT-TYPE>_FILE_<DEP>_<NAME>` for each artifact by name (e.g.
+`CARGO_BIN_FILE_EXAMPLE_mybin`). Note that the name you give to the dependency
+determines the `<DEP>`, but does not affect the `<NAME>` of each artifact
+within that dependency.
+
+Cargo will unify features and versions across all kinds of dependencies,
+including multiple artifact dependencies, just as it does for multiple
+dependencies on the same crate throughout a dependency tree. A dependency tree
+may only include one semver-compatible version of a given crate, but may
+include multiple semver-incompatible versions of a given crate.
+
+Building an artifact dependency for multiple targets may entail building
+multiple copies of other dependencies, which must similarly unify within a
+dependency tree.
+
+Multiple dependencies on the same crate may specify different values for
+`artifact` (e.g. to build a library and/or multiple specific binaries), as well
+as different values for `target`. Cargo will combine all the entries for a
+given `target`, and build all the specified artifacts for that target.
+Requesting a specific artifact for one target will not affect the artifacts
+built for another target.
+
+[Profile
+overrides](https://doc.rust-lang.org/cargo/reference/profiles.html#overrides)
+are specified in terms of the original crate name, not the dependency name;
+thus, Cargo does not currently support overriding profile settings differently
+for different artifact dependencies.
+
+Until this feature is stabilized, it will require specifying the nightly-only
+option `-Z multidep` to `cargo`. If `cargo` encounters multiple dependencies on
```

Comment:
As far as I can tell, some things in `Cargo.toml` have `-Z` flags; for instance, you added ``-Z weak-dep-features`, which controls an addition to the syntax in `Cargo.toml`. `-Z namespaced-features` seems similar.

I don't have a preference here; just trying to follow the example of existing features.

Reply:
Ah, those use `-Z` because it affects the resolver which is a global thing (if a dependency used the new syntax, it wouldn't have been possible to determine that ahead of time).  I guess it depends on whether or not this affects the index.  For some reason I was thinking that these would be combined ahead of time before adding to the index, but thinking about it some more that's probably not a good idea.  So a `-Z` flag is probably appropriate.

We may need to figure out some way to prevent these from being published (I'm not sure if crates.io will check).

Reply:
> We may need to figure out some way to prevent these from being published (I'm not sure if crates.io will check).

Permanently or until stabilization? If this doesn't break the index for existing cargo versions, it shouldn't prevent publishing at all IMO. If it does break the index for existing cargo versions, it should be permanently prevented to ensure that older cargo versions keep working in the future after stabilization of this feature.

Reply:
I also don't think it is necessary to affect the index. The duplicate dependencies could be listed as a single dependency in the index. That should give enough information for the resolver anyway.

Reply:
Until stabilized.  In general, we don't want to allow publishing unstable changes to the index until we are confident that the design is correct since the index is permanent.

Yea, I'm a little uncertain if this should be combined before adding to the index or not.  I think not, since internally cargo works on the index data.

I did a quick test, and it doesn't appear to break the index if they aren't combined.  When selecting a version that contains these entries, the resolver gives an error.  We could possibly bump the `v` field to force it to skip those entries.

Regardless, I don't think this needs to be determined now. These kind of details can be explored during implementation.